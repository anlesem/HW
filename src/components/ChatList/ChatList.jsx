import { useEffect, useRef, useState } from 'react';
import { NavLink } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

import { getChatList, getChatCounter } from '../../store/chats/selectors';
import { addChat, renameChat, deleteChat } from '../../store/chats/actions';
import { initMessageList, initTempInput } from '../../store/messages/actions';

import style from './ChatList.module.scss';

import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import Checkbox from '@mui/material/Checkbox';
import IconButton from '@mui/material/IconButton';
import CommentIcon from '@mui/icons-material/Comment';

import Stack from '@mui/material/Stack';
import Button from '@mui/material/Button';
import AddIcon from '@mui/icons-material/Add';
import DeleteIcon from '@mui/icons-material/Delete';

import SendIcon from '@mui/icons-material/Send';
import TextField from '@mui/material/TextField';

export const ChatList = () => {
  const dispatch = useDispatch();

  const chats = useSelector(getChatList);
  const counter = useSelector(getChatCounter);
  const [text, setText] = useState('');
  const [checked, setChecked] = useState([0]);
  const scrollChats = useRef(null);

  const handleToggle = (value) => () => {
    const currentIndex = checked.indexOf(value);
    const newChecked = [...checked];
    if (currentIndex === -1) {
      newChecked.push(value);
    } else {
      newChecked.splice(currentIndex, 1);
    }
    setChecked(newChecked);
  };

  const handleAdd = () => {
    dispatch(addChat);
    dispatch(initMessageList(counter + 1));
    dispatch(initTempInput);
    scrollChats.current.scrollTop =
      scrollChats.current.scrollHeight - scrollChats.current.clientHeight;
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    if (text) {
      dispatch(renameChat(checked[1], text));
      setText('');
      setChecked([0]);
    }
  };

  const handleDelete = (checked) => {
    checked.forEach((id) => {
      dispatch(deleteChat(id));
    });
    setChecked([0]);
  };

  const handleInput = (value) => {
    setText(value);
  };

  const handleKeyDown = (event) => {
    if (event.keyCode === 13) {
      event.preventDefault();
      handleSubmit(event);
    }
  };

  useEffect(() => {
    setText('');
  }, [checked]);

  return (
    <div className={style.wrap}>
      <List
        sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}
        className={style.chatList}
        ref={scrollChats}>
        {chats.map((chat) => {
          const labelId = `checkbox-list-label-${chat.id}`;
          return (
            <ListItem
              key={chat.id}
              secondaryAction={
                <NavLink
                  to={`/chats/${chat.id}`}
                  className={({ isActive }) => (isActive ? style.active : '')}>
                  <IconButton
                    edge="end"
                    aria-label="comments"
                    data-testid={'chat-change-' + chat.id}>
                    <CommentIcon />
                  </IconButton>
                </NavLink>
              }
              disablePadding>
              <ListItemButton
                role={undefined}
                onClick={handleToggle(chat.id)}
                dense
                data-testid={'chat-item-' + chat.id}>
                <ListItemIcon>
                  <Checkbox
                    edge="start"
                    checked={checked.indexOf(chat.id) !== -1}
                    tabIndex={-1}
                    disableRipple
                    inputProps={{
                      'aria-labelledby': labelId,
                      'data-testid': `chat-check-${chat.id}`
                    }}
                  />
                </ListItemIcon>
                <ListItemText id={labelId} primary={chat.name} />
              </ListItemButton>
            </ListItem>
          );
        })}
      </List>
      <Stack direction="row" spacing={2} className={style.manage}>
        <Button
          variant="outlined"
          startIcon={<DeleteIcon />}
          disabled={checked.length > 1 ? false : true}
          data-testid="chats-button-remove"
          onClick={() => handleDelete(checked)}>
          Удалить
        </Button>
        <Button
          variant="outlined"
          startIcon={<AddIcon />}
          data-testid="chats-button-add"
          onClick={() => handleAdd(checked)}>
          Добавить
        </Button>
      </Stack>
      {checked.length === 2 && (
        <form className={style.rename} onSubmit={handleSubmit}>
          <TextField
            id="outlined-basic"
            name="chat"
            label="Название чата"
            variant="outlined"
            value={text}
            onChange={(event) => handleInput(event.target.value)}
            onKeyDown={(event) => handleKeyDown(event)}
            required
            inputProps={{ 'data-testid': 'chat-form-input' }}
            className={style.textField}
          />
          <Button variant="outlined" type="submit" data-testid="chat-form-button">
            <SendIcon />
          </Button>
        </form>
      )}
    </div>
  );
};
