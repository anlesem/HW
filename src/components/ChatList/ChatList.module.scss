.chatList {
   width: 35%;
   height: 100%;
   padding: 20px 20px 50px;
   font-size: 20px;
   font-weight: semibold;

   overflow-y: auto;
}

.form {
   position: absolute;
   bottom: 20px;
   background-color: #fff;
}

.rename {
   width: 100%;
   background-color: #fff;
   margin-bottom: 20px;

   display: flex;

   button {
      margin-left: 10px;
   }
}

.manage {
   width: 100%;
   background-color: #fff;

   display: flex;
   justify-content: space-between;
}

.active {

   button {
      border-radius: 50%;
      background-color: #555;
   }

   svg {
      fill: #fff;
   }
}