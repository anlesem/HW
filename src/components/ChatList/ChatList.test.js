import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { ChatList } from './ChatList';

describe('ChatList', () => {
  it('Компонент существует', () => {
    expect(ChatList).toBeInstanceOf(Function);
  });

  it('Снимок состояния', () => {
    const handleToggle = jest.fn();
    const chats = [
      {
        id: 1,
        name: 'чат 1'
      }
    ];
    const checked = [];
    const scrollChats = null;

    const { asFragment } = render(
      <BrowserRouter>
        <ChatList data={{ handleToggle, chats, checked, scrollChats }} />
      </BrowserRouter>
    );
    expect(asFragment(<ChatList />)).toMatchSnapshot();
  });

  it('Сформированный список чатов', () => {
    const handleToggle = jest.fn();
    const chats = [
      {
        id: 1,
        name: 'чат 1'
      }
    ];
    const checked = [];
    const scrollChats = null;

    render(
      <BrowserRouter>
        <ChatList data={{ handleToggle, chats, checked, scrollChats }} />
      </BrowserRouter>
    );

    expect(screen.getByRole('listitem')).toBeTruthy();
    expect(screen.getByText(/чат/)).toBeTruthy();
  });
});
