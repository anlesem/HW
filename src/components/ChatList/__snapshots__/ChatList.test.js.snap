// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatList Снимок состояния 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root MuiList-padding chatList css-1mcnwpj-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters css-g28sfz-MuiListItem-root"
    >
      <div
        class="MuiListItemButton-root MuiListItemButton-dense MuiListItemButton-gutters MuiButtonBase-root css-1m4u3pv-MuiButtonBase-root-MuiListItemButton-root"
        data-testid="chat-item-1"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root PrivateSwitchBase-edgeStart css-zkqepn-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              aria-labelledby="checkbox-list-label-1"
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              data-testid="chat-check-1"
              tabindex="-1"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          id="checkbox-list-label-1"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
          >
            чат 1
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
      <div
        class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
      >
        <a
          class=""
          href="/chats/1"
        >
          <button
            aria-label="comments"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            data-testid="chat-change-1"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CommentIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M21.99 4c0-1.1-.89-2-1.99-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4-.01-18zM18 14H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
